pyccel/parser/semantic.py

```diff
@@ -3608,13 +3621,35 @@ def _visit_ClassDef(self, expr):
         for i in methods:
             self._visit_FunctionDef(i)
 
+        if not any(method.name == '__del__' for method in methods):
+            argument = FunctionDefArgument(Variable(cls.name, 'self', cls_base = cls))
+            scope = self.create_new_function_scope('__del__')
+            del_method = FunctionDef('__del__', [argument], (), [Pass()], cls_name=cls.name, scope=scope)
+            self.exit_function_scope()
+            self.insert_function(del_method)
+            cls.add_new_method(del_method)
+
+        for method in cls.methods:
+            if method.name == '__del__':
+                self._current_function = method.name
+                attribute = [attr for attr in cls.attributes if not attr.on_stack]
+                if attribute:
+                    self._allocs[-1].extend(attribute)
+                    is_freed = next(is_freed for is_freed in cls.attributes if is_freed.name.startswith("is_freed"))
```

Comment:
You should still have access to `deallocater_rhs` here. You can use it directly instead of searching for it. This will avoid any potential problems due to name collisions

Reply:
Fixed https://github.com/pyccel/pyccel/pull/1508/commits/fe0041d898f4fc1f33be35c56f9532bf83f884f7.
The condition was in the beggining just if their is some attributes of the class that should be deleted.
But now the check is on all the condition even if the `__del__` function is empty.